packages/viewer/src/modules/UrlHelper.ts

```diff
@@ -0,0 +1,276 @@
+    },
+    ref: ReferencedObjectUrl
+  ): Promise<string[]> {
+    return this.runAllModelsQuery(headers, ref)
+  }
+
+  private static async runModelLastVersionQuery(
+    headers: { 'Content-Type': string; Authorization: string },
+    ref: ReferencedObjectUrl,
+    resource: SpeckleViewer.ViewerRoute.ViewerModelResource
+  ): Promise<string> {
+    const res = await fetch(`${ref.origin}/graphql`, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify({
+        query: `
+          query ExampleQuery($modelId: String!, $projectId: String!) {
+            project(id: $projectId) {
+              model(id: $modelId) {
+                versions(limit: 1) {
+                  items {
+                    referencedObject
+                  }
+                }
+              }
+            }
+          }
+        `,
+        variables: {
+          projectId: ref.projectId,
+          modelId: resource.modelId
+        }
+      })
+    })
+    const { data } = await res.json()
+    return `${ref.origin}/streams/${ref.projectId}/objects/${data.project.model.versions.items[0].referencedObject}`
+  }
+
+  private static async runModelVersionQuery(
+    headers: { 'Content-Type': string; Authorization: string },
+    ref: ReferencedObjectUrl,
+    resource: SpeckleViewer.ViewerRoute.ViewerModelResource
+  ): Promise<string> {
+    const res = await fetch(`${ref.origin}/graphql`, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify({
+        query: `
+          query ExampleQuery($modelId: String!, $projectId: String!, $versionId: String!) {
+            project(id: $projectId) {
+              model(id: $modelId) {
+                version(id: $versionId) {
+                  referencedObject
+                }
+              }
+            }
+          }
+        `,
+        variables: {
+          projectId: ref.projectId,
+          modelId: resource.modelId,
+          versionId: resource.versionId
+        }
+      })
+    })
+    const { data } = await res.json()
+    return `${ref.origin}/streams/${ref.projectId}/objects/${data.project.model.version.referencedObject}`
+  }
+
+  private static async runAllModelsQuery(
+    headers: { 'Content-Type': string; Authorization: string },
+    ref: ReferencedObjectUrl
+  ): Promise<string[]> {
+    const res = await fetch(`${ref.origin}/graphql`, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify({
+        query: `
+         query Query($projectId: String!) {
+          project(id: $projectId) {
+            models {
+              items {
+                versions(limit: 1) {
+                  items {
+                    referencedObject
+                  }
+                }
+              }
+            }
+          }
+        }
+        `,
+        variables: {
+          projectId: ref.projectId
+        }
+      })
+    })
+    const { data } = await res.json()
+    const urls: string[] = []
+    data.project.models.items.forEach(
```

Comment:
`data` can easily be undefined if there's an error during the execution of the query. Check for this in all queries.

Reply:
`project` can be null too, if its not found